File: LeetCode/iter1/c++/388.longest-absolute-file-path.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <queue>
#include <stack>

using namespace std;

class Solution {
public:
	int lengthLongestPath(string input) {
		vector<int> dirStack;
		string tmpPathStr = "";
		int ret = 0;
		int curLen = 0;
		int preIdx = 0;
		input += "\n";
		while (preIdx < input.length()) {
			//截取\n之间的子串
			int curIdx = input.find("\n", preIdx);
			tmpPathStr = input.substr(preIdx, curIdx - preIdx);
			preIdx = curIdx + 1;

			//获取当前路径层级，并回溯
			int level = tmpPathStr.find_last_of("\t");
			level = level == string::npos ? 0 : level + 1;
			while (dirStack.size() > level) {
				curLen -= dirStack.back();
				dirStack.pop_back();
			}

			// eg:/file.txt
			int pathLen = tmpPathStr.size() - level + 1;
			curLen += pathLen;
			dirStack.push_back(pathLen);

			if (tmpPathStr.find(".") != string::npos) {
				ret = max(curLen - 1, ret); //去除开头的/
			}
		}

		return ret;
	}
};

int main()
{
	Solution solution;
	cout << solution.lengthLongestPath("dir\n\tsubdir1\n\tsubdir2\n\t\tfile.ext") << endl;
	cout << solution.lengthLongestPath("dir\n\tsubdir1\n\t\tfile1.ext\n\t\tsubsubdir1\n\tsubdir2\n\t\tsubsubdir2\n\t\t\tfile2.ext") << endl;

	return 0;
}

File: LeetCode/iter1/java/388.longest-absolute-file-path.java
package org.ironman.base.playground.leetcode;

import java.util.Stack;

public class Solution {

    public static void main(String[] args) {
        System.out.println(lengthLongestPath("dir\n\tsubdir1\n\tsubdir2\n\t\tfile.ext"));
        System.out.println(lengthLongestPath("dir\n\tsubdir1\n\t\tfile1.ext\n\t\tsubsubdir1\n\tsubdir2\n\t\tsubsubdir2\n\t\t\tfile2.ext"));
    }

    public static int lengthLongestPath(String input) {
        String[] path = input.split("\n");
        Stack<Integer> stack = new Stack<>();
        int curLen = 0;
        int ret = 0;

        for (String s : path) {
            int level = s.lastIndexOf("\t") + 1;

            while (stack.size() > level) {
                curLen -= stack.pop();
            }

            //eg: /file.txt
            int pathLen = "/".length() + s.length() - level;
            curLen += pathLen;

            if (s.contains(".")) {
                ret = Math.max(ret, curLen - 1); //starter slash
            }

            stack.push(pathLen);
        }

        return ret;
    }
}